### PR TITLE
TC-1

### PR DESCRIPTION
**Pull request by Lou-Kamades**

### CODE REVIEW

File: bench/src/benches/api_load.rs [hunk 1]
```diff
@@ -13,7 +13,12 @@ use solana_sdk::signature::{read_keypair_file, Keypair, Signer};
 use crate::create_memo_tx_small;
 
 // TC3 measure how much load the API endpoint can take
-pub async fn api_load(payer_path: &Path, rpc_url: String, time_ms: u64) -> anyhow::Result<()> {
+pub async fn api_load(
+    payer_path: &Path,
+    rpc_url: String,
+    time_ms: u64,
```
Comment (grooviegermanikus) [inline bench/src/benches/api_load.rs:19]: maybe rename to clarify what that parameter does

File: bench/src/benches/confirmation_slot.rs [hunk 2]
```diff
@@ -1,43 +1,107 @@
...
+use solana_rpc_client_api::config::{RpcSendTransactionConfig, RpcTransactionConfig};
+use solana_sdk::signature::{read_keypair_file, Signature, Signer};
 use solana_sdk::transaction::Transaction;
 use solana_sdk::{commitment_config::CommitmentConfig, signature::Keypair};
+use solana_transaction_status::UiTransactionEncoding;
+use tokio::time::{sleep, Instant};
+use tracing::error;
 
-/// TC1 send 2 txs (one via LiteRPC, one via Solana RPC) and compare confirmation slot (=slot distance)
+#[derive(Default, PartialEq, Eq)]
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:17]: could PartialEq+Eq be removed. looks wrong to me

File: bench/src/benches/confirmation_rate.rs [hunk 3]
```diff
@@ -62,10 +70,18 @@ pub async fn send_bulk_txs_and_wait(
...
     let hash = rpc.get_latest_blockhash().await?;
     let mut rng = create_rng(None);
-    let txs = generate_txs(num_txns, payer, hash, &mut rng, tx_size);
+    let txs = generate_txs(
+        num_txns,
+        payer,
+        hash,
+        &mut rng,
+        tx_size,
+        cu_price_micro_lamports,
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_rate.rs:83]: we could bundle parameters for tx creation in a struct:
* tx_size
* cu_price_micro_lamports

File: bench/src/benches/confirmation_slot.rs [hunk 4]
```diff
@@ -47,19 +111,99 @@ async fn create_tx(
...
 
-async fn send_transaction_and_get_slot(client: &RpcClient, tx: Transaction) -> anyhow::Result<u64> {
-    let status = send_and_confirm_transactions(client, &[tx], CommitmentConfig::confirmed(), None)
-        .await?
-        .into_iter()
-        .next()
-        .unwrap()?
-        .context("unable to confirm tx")?;
+pub async fn rpc_roundtrip_duration(rpc: &RpcClient) -> anyhow::Result<Duration> {
+    let start = Instant::now();
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:181]: nit: I use `started_at` consistently for such variables

File: bench/src/benches/confirmation_slot.rs [hunk 5]
```diff
@@ -1,43 +1,107 @@
...
     let mut rng = create_rng(None);
     let payer = read_keypair_file(payer_path).expect("payer file");
     info!("Payer: {}", payer.pubkey().to_string());
 
-    let rpc_a_tx = create_tx(&rpc_a, &payer, &mut rng, tx_size).await?;
-    let rpc_b_tx = create_tx(&rpc_b, &payer, &mut rng, tx_size).await?;
+    for _ in 0..num_rounds {
+        let rpc_a = RpcClient::new(rpc_a_url.clone());
+        let rpc_b = RpcClient::new(rpc_b_url.clone());
+        let time_a = rpc_roundtrip_duration(&rpc_a).await?.as_secs_f64();
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:51]: could add a comment like ...

> measure network time to reach the respective RPC endpoints to later mitigate the distance by delaying the call

File: bench/src/benches/confirmation_slot.rs [hunk 6]
```diff
@@ -1,43 +1,107 @@
...
 pub async fn confirmation_slot(
     payer_path: &Path,
     rpc_a_url: String,
     rpc_b_url: String,
     tx_size: TxSize,
+    max_timeout_ms: u64,
+    num_rounds: usize,
+    cu_price_micro_lamports: u64,
+    ping_thing_token: Option<String>,
 ) -> anyhow::Result<()> {
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:39]: we need to design a metric result struct - right?

**Review by grooviegermanikus — CHANGES_REQUESTED** [review]

File: bench/src/benches/confirmation_slot.rs [hunk 7]
```diff
@@ -1,43 +1,107 @@
...
+                })
+        });
+
+        let b_task = tokio::spawn(async move {
+            sleep(Duration::from_secs_f64(b_delay)).await;
+            send_transaction(&rpc_b, rpc_b_tx, max_timeout_ms)
+                .await
+                .unwrap_or_else(|e| {
+                    error!("Failed to confirm txn for B: {}", e);
+                    ConfirmationSlotResult::default()
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:87]: IMO this should be `Result<ConfirmationSlotResult>`

File: bench/src/lib.rs [hunk 8]
```diff
@@ -199,36 +200,51 @@ pub fn create_memo_tx(
...
 
-    // TODO make configurable
-    // 3 -> 6 slots
-    // 1 -> 31 slots
-    let cu_budget: Instruction = ComputeBudgetInstruction::set_compute_unit_price(3);
+    let cu_budget_ix: Instruction =
+        ComputeBudgetInstruction::set_compute_unit_price(cu_price_micro_lamports);
     // Program consumed: 12775 of 13700 compute units
-    let priority_fees: Instruction = ComputeBudgetInstruction::set_compute_unit_limit(14000);
+    let cu_limit_ix: Instruction = ComputeBudgetInstruction::set_compute_unit_limit(14000);
```
Comment (grooviegermanikus) [inline bench/src/lib.rs:228]: could you port the cu_requested over from https://github.com/blockworks-foundation/lite-rpc/pull/363/commits/73d8823b28a684fecad2be1b8b41d60614fc9b83

File: bench/src/benches/confirmation_slot.rs (hunk 6 above)
Comment (Lou-Kamades) [reply bench/src/benches/confirmation_slot.rs:57]: going to put this in a different PR to keep things manageable but I was thinking:

CREATE TABLE benchmarks.confirmation_slot (
	signature text NOT NULL,
	bench_datetime timestamptz NOT NULL,
	slot_sent int8 NOT NULL,
	slot_landed int8,
	endpoint text not null,
	confirmed bool not null,
	confirmation_time_ms int4,
	CONSTRAINT signature_slot_pkey PRIMARY KEY (signature, slot)
);

File: bench/src/benches/confirmation_slot.rs [hunk 9]
```diff
@@ -46,20 +155,70 @@ async fn create_tx(
...
+
+async fn send_and_confirm_transaction(
+    rpc: &RpcClient,
+    tx: Transaction,
+) -> anyhow::Result<ConfirmationSlotInfo> {
+    let result_vec: Vec<(Signature, ConfirmationResponseFromRpc)> =
+        send_and_confirm_bulk_transactions(rpc, &[tx]).await?;
 
-    Ok(create_memo_tx(payer, hash, rng, tx_size))
+    let (signature, confirmation_response) = result_vec.into_iter().next().unwrap();
```
Comment (grooviegermanikus) [inline bench/src/benches/confirmation_slot.rs:174]: blocker: we need to build a better result struct here

**Pull request merged**